lib/rules/prefer-named-capture-group.js

```diff
@@ -68,12 +74,35 @@ module.exports = {
             regexpp.visitRegExpAST(ast, {
                 onCapturingGroupEnter(group) {
                     if (!group.name) {
+                        const rawText = sourceCode.getText(regexNode);
+                        const start = regexNode.range[0] + rawText.indexOf("(", group.start) + 1;
```

Comment:
This won't always find the correct `start`. For example:

```js
new RegExp("\u1234\u5678(?:a)(b)");
```

```
Capture group '(b)' should be converted to a named or non-capturing group.
```

"Add name to capture group" suggestion adds name to `(?:a)` instead of `(b)`:

```js
new RegExp("\u1234\u5678(?<temp1>?:a)(b)");
```

We could check if `regexNode` is a regex/string/template literal with raw text (the part that represents the pattern, e.g.., source code text between quotes if it's a string literal) that is the same as `pattern`, and only in that case provide suggestions.

Reply:
Err, I think it does, because it passes `group.start` as the starting search index. I added a test for this specific cause in c57ed1a9dde771e8458c6d979d9668dc205d48c0.

Reply:
Err, does it? I added a test in c57ed1a9dde771e8458c6d979d9668dc205d48c0 that shows the correct suggestion behavior.

I would think the passing of `group.start` to the `indexOf()` would fix for this, no?

Reply:
I left https://github.com/eslint/eslint/pull/16544#discussion_r1022535262 on the test.

Reply:
The problem is that the pattern we're passing to regexpp is string literal's value, and all indexes that regexpp calculates (e.g., `group.start`), are indexes in that value.

In the above example, we have a string literal that has 20 source code characters between quotes:

```
\u1234\u5678(?:a)(b)
```

but, regexpp works with a string that has 10 characters:

```
ሴ噸(?:a)(b)
```

so, when we try to fix the string literal based on indexes from its value, the result can be wrong.

 I think it's fine to provide suggestions only if the string literal's source code between quotes is the same as its value, since that's the only case where we already have the correct indexes (as they're same). Otherwise, we would need to map indexes, but using `new RegExp` with string literals should be rare (regex literals should be used instead of `new RegExp`), so it doesn't seem worth adding the complexity or adding a new dependency if there's already a library for this.

Reply:
Aha, thanks for the explanation - this makes sense.

Do you think it's enough to check for the presence of `\\` in the raw text to determine whether to provide suggestions? I don't know of other ways to cause index discrepencies.

Reply:
It depends on what the `regexNode` is. First, this rule is using [`getStringIfConstant`](https://github.com/eslint/eslint/blob/436f712843360f98b2bd63256bf0c4f77013b54c/lib/rules/prefer-named-capture-group.js#L100), which can calculate expressions so `regexNode` can be complex. For example:

```js
new RegExp("foo" + "(?:a)" + "(b)");
```

```
Capture group '(b)' should be converted to a named or non-capturing group.
```

The current iteration of this PR suggests a fix that adds a name to `(?:a)` instead of `(b)`:


```js
new RegExp("foo" + "(?<temp1>?:a)" + "(b)");
```

We should first determine what kind of node `regexNode` is:

* If it's a regex literal, then we can always provide suggestions. I think there's no need to check the raw pattern, as it should be the same as the evaluated pattern.
* If it's a string literal, checking for the presence of `\\` looks good. Alternatively, we could check if `rawText.slice(1, -1) === pattern` to be sure.
* If it's a template literal with no expressions, checking `rawText.slice(1, -1) === pattern` seems the safest. 
* Else, it's a complex expression so we can't provide suggestions.

Reply:
Sorry, I missed that this was a new test case - added!